pkg/datapath/linux/config/config.go

```diff
@@ -224,14 +224,18 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		cDefinesMap["ENABLE_PREFILTER"] = "1"
 	}
 
+	if option.Config.BPFSocketLBHostnsOnly {
+		cDefinesMap["ENABLE_HOST_XLATE_ONLY"] = "1"
+	}
+
 	if option.Config.EnableHostReachableServices {
 		if option.Config.EnableHostServicesTCP {
 			cDefinesMap["ENABLE_HOST_SERVICES_TCP"] = "1"
 		}
 		if option.Config.EnableHostServicesUDP {
 			cDefinesMap["ENABLE_HOST_SERVICES_UDP"] = "1"
 		}
-		if option.Config.EnableHostServicesTCP && option.Config.EnableHostServicesUDP {
+		if option.Config.EnableHostServicesTCP && option.Config.EnableHostServicesUDP && !option.Config.BPFSocketLBHostnsOnly {
 			cDefinesMap["ENABLE_HOST_SERVICES_FULL"] = "1"
```

Comment:
I'm guessing it is this change that is causing complexity issues, because we now enable a bunch of additional code (guarded by `#ifndef ENABLE_HOST_SERVICES_FULL` in `bpf_lxc`) when `option.Config.EnableHostServices{TCP,UDP}=true` but `option.Config.BPFSocketLBHostnsOnly=true`.

Reply:
Your pull request did not introduce this complexity issue. It already happens when compiling both NodePort handling and full host service translation in the BPF datapath. I'm able to reproduce in the development VM with the following config:
```diff
diff --git a/contrib/vagrant/start.sh b/contrib/vagrant/start.sh
index 9f9c22ab0..767929f60 100755
--- a/contrib/vagrant/start.sh
+++ b/contrib/vagrant/start.sh
@@ -306,7 +306,7 @@ function write_cilium_cfg() {
     ipv6_addr="${3}"
     filename="${4}"
 
-    cilium_options=" --debug --pprof --enable-hubble --hubble-listen-address :4244 --enable-k8s-event-handover --k8s-require-ipv4-pod-cidr --auto-direct-node-routes"
+    cilium_options=" --debug --pprof --enable-hubble --hubble-listen-address :4244 --enable-k8s-event-handover --k8s-require-ipv4-pod-cidr --auto-direct-node-routes --kube-proxy-replacement=partial --devices=enp0s3,enp0s8 --host-reachable-services-protos=tcp --enable-node-port --enable-external-ips --enable-host-port --enable-session-affinity"
     cilium_operator_options=" --debug"
 
     if [[ "${IPV4}" -eq "1" ]]; then
```
It's basically `kube-proxy-replacement=strict` but with `host-reachable-services-protos=tcp` instead of `tcp,udp`.

That configuration isn't tested as part of our CI today :disappointed:

Reply:
Discussion moved to #14234.